kubernetes/websocket/wsclient.c

```diff
@@ -0,0 +1,337 @@
+{
+    if (!base_path) {
+        fprintf(stderr, "%s: The base path is invalid.\n", __func__);
+        return NULL;
+    }
+
+    wsclient_t *wsc = (wsclient_t *) calloc(1, sizeof(wsclient_t));
+    if (!wsc) {
+        fprintf(stderr, "%s: Cannot allocate the memory for websocket client.\n", __func__);
+        return NULL;
+    }
+
+    int rc = get_server_address_and_port(&wsc->server_address, &wsc->server_port, base_path);
+    if (0 != rc) {
+        fprintf(stderr, "%s: Cannot get the websocket server address or port from the base path <%s>.\n", __func__, base_path);
+        goto error;
+    }
+
+    wsc->ssl_config = ssl_config;
+    wsc->log_mask = ws_log_mask;
+
+    return wsc;
+
+error:
+    if (wsc) {
+        free(wsc);
+        wsc = NULL;
+    }
+    return NULL;
+}
+
+void wsclient_free(wsclient_t * wsc)
+{
+    if (!wsc) {
+        return;
+    }
+    if (wsc->server_address) {
+        free(wsc->server_address);
+        wsc->server_address = NULL;
+    }
+    if (wsc->data_received) {
+        free(wsc->data_received);
+        wsc->data_received = NULL;
+    }
+    wsc->data_received_len = 0;
+    wsc->data_callback_func = NULL;
+    wsc->ssl_config = NULL;
+
+    free(wsc);
+}
+
+/*
+ * Scheduled sul callback that starts the connection attempt
+ */
+
+static void connect_client(lws_sorted_usec_list_t * sul)
+{
+    wsclient_t *wsc = lws_container_of(sul, wsclient_t, sul);
+    struct lws_client_connect_info i;
+
+    memset(&i, 0, sizeof(i));
+
+    i.context = g_lws_context;
+    i.port = wsc->server_port;
+    i.address = wsc->server_address;
+    i.path = wsc->path;
+    i.host = i.address;
+    i.origin = i.address;
+    i.ssl_connection = wsc->ssl_config ? LCCSCF_USE_SSL : 0;
+    
+    //i.protocol = pro;
+    //i.local_protocol_name = "websocket-client";
+    i.pwsi = &wsc->wsi;
+    i.retry_and_idle_policy = &retry;
+    i.userdata = wsc;
+
+    if (!lws_client_connect_via_info(&i))
+        /*
+         * Failed... schedule a retry... we can't use the _retry_wsi()
+         * convenience wrapper api here because no valid wsi at this
+         * point.
+         */
+        if (lws_retry_sul_schedule(g_lws_context, 0, sul, &retry, connect_client, &wsc->retry_count)) {
+            lwsl_err("%s: connection attempts exhausted\n", __func__);
+            g_interrupted = true;
+        }
+}
+
+static int callback_wsclient(struct lws *wsi, enum lws_callback_reasons reason, void *user, void *in, size_t len)
+{
+    lwsl_user("%s: callback reason %d\n", __func__, reason);
+
+    wsclient_t *wsc = (wsclient_t *) user;
+
+    switch (reason) {
+
+    case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
+        lwsl_err("%s: wsclient connection error: %s\n", __func__, in ? (char *) in : "(null)");
+        goto do_retry;
+        break;
+
+    case LWS_CALLBACK_CLIENT_ESTABLISHED:
+        lwsl_user("%s: wsclient connection established\n", __func__);
+        lws_callback_on_writable(wsi);
+        break;
+
```

Comment:
for the incluster_config
I add the this case for adding token to header
    case LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER:
    {
        const char *tpath = "/var/run/secrets/kubernetes.io/serviceaccount/token";
        char *token = read_complete_file(tpath);
        char *token_header = strcat_copy("Bearer ", token);
        const char *acc = "*/*";

        unsigned char **p = (unsigned char **)in, *end = (*p) + len;

        if (lws_add_http_header_by_token(wsi, WSI_TOKEN_HTTP_ACCEPT,
                (unsigned char *)acc, (int)strlen(acc), p, end))
        {
            free(token_header);
            return -1;
        }

        if (lws_add_http_header_by_token(wsi, WSI_TOKEN_HTTP_AUTHORIZATION,
                (unsigned char *)token_header, (int)strlen(token_header), p, end))
        {
            free(token_header);
            return -1;
        }

        free(token_header);
        break;
    }

Reply:
Thanks. It's helpful.